File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/repository/CloseableNonExclusiveTransaction.java
package com.lasthopesoftware.bluewater.repository;

import android.database.sqlite.SQLiteDatabase;

import java.io.Closeable;

public class CloseableNonExclusiveTransaction implements Closeable, ITransactionSuccessSetter {

    private final SQLiteDatabase sqLiteDatabase;

    CloseableNonExclusiveTransaction(SQLiteDatabase sqLiteDatabase) {
        this.sqLiteDatabase = sqLiteDatabase;
        this.sqLiteDatabase.beginTransactionNonExclusive();
    }

    public void setTransactionSuccessful() {
        this.sqLiteDatabase.setTransactionSuccessful();
    }

    @Override
    public void close() {
        this.sqLiteDatabase.endTransaction();
    }
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/repository/CloseableNonExclusiveTransaction.kt
package com.lasthopesoftware.bluewater.repository

import android.database.sqlite.SQLiteDatabase
import java.io.Closeable
import java.util.concurrent.locks.ReadWriteLock

class CloseableNonExclusiveTransaction internal constructor(private val sqLiteDatabase: SQLiteDatabase, databaseSynchronization: ReadWriteLock) : Closeable, ITransactionSuccessSetter {

	private val readLock = databaseSynchronization.readLock()

	init {
		readLock.lock()
		sqLiteDatabase.beginTransactionNonExclusive()
	}

	override fun setTransactionSuccessful() {
		sqLiteDatabase.setTransactionSuccessful()
	}

	override fun close() {
		sqLiteDatabase.endTransaction()
		readLock.unlock()
	}
}
